checker/src/org/checkerframework/checker/upperbound/UpperBoundVisitor.java

```diff
@@ -0,0 +1,131 @@
+     * MinLenATF. If the MinLen of the array is known, the visitor can check if the index is less
+     * than the MinLen, using the Value Checker. If so then the access is still safe. Otherwise,
+     * report a potential unsafe access.
+     */
+    @Override
+    public Void visitArrayAccess(ArrayAccessTree tree, Void type) {
+        ExpressionTree indexTree = tree.getIndex();
+        ExpressionTree arrTree = tree.getExpression();
+        String arrName = FlowExpressions.internalReprOf(this.atypeFactory, arrTree).toString();
+        AnnotatedTypeMirror indexType = atypeFactory.getAnnotatedType(indexTree);
+
+        // Need to be able to check these as part of the conditional below.
+        // Find max because it's important to determine whether the index is
+        // less than the minimum length of the array. If it could be any
+        // of several values, only the max is of interest.
+        Integer valMax = atypeFactory.valMaxFromExpressionTree(indexTree);
+        Integer minLen = atypeFactory.minLenFromExpressionTree(arrTree);
+
+        // Is indexType LTL/LTOM of a set containing arrName?
+        if ((indexType.hasAnnotation(LTLengthOf.class)
+                        || indexType.hasAnnotation(LTOMLengthOf.class))
+                && (UpperBoundUtils.hasValue(indexType, arrName)
+                        || UpperBoundUtils.hasValue(indexType, arrTree.toString()))) {
+            // If so, this is safe - get out of here.
+            return super.visitArrayAccess(tree, type);
+        } else if (valMax != null && minLen != null && valMax < minLen) {
+            return super.visitArrayAccess(tree, type);
+        } else {
+            // Unsafe, since neither the Upper bound or MinLen checks succeeded.
+            checker.report(Result.warning(UPPER_BOUND, indexType.toString(), arrName), indexTree);
+            return super.visitArrayAccess(tree, type);
+        }
+    }
+
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree tree, Void type) {
+        if (isFirstArgListMethod(tree)) {
+            ExpressionTree indexTree = tree.getArguments().get(0);
+            ExpressionTree lstTree = tree.getMethodSelect();
+            String lstName = FlowExpressions.internalReprOf(this.atypeFactory, lstTree).toString();
+            AnnotatedTypeMirror indexType = atypeFactory.getAnnotatedType(indexTree);
+            String[] parts = lstTree.toString().split("\\.");
+            // the last part of the methodName is the method (get, remove, etc) so get second to last part
+            String localName = parts[parts.length - 2];
+            // Need to be able to check these as part of the conditional below.
+            // Find max because it's important to determine whether the index is
+            // less than the minimum length of the array. If it could be any
+            // of several values, only the max is of interest.
+            Integer valMax = atypeFactory.valMaxFromExpressionTree(indexTree);
```

Comment:
This code appears to be copied from the array index check above.  Please put the duplicated code in a method and call it in both places.